trillian/migrillian/election/etcd/election.go

```diff
@@ -0,0 +1,152 @@
+// Copyright 2018 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package etcd provides implementation master election based on etcd.
+package etcd
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/coreos/etcd/clientv3"
+	"github.com/coreos/etcd/clientv3/concurrency"
+	"github.com/golang/glog"
+	"github.com/google/certificate-transparency-go/trillian/migrillian/election"
+)
+
+// Election is an implementation of election.Election based on etcd.
+type Election struct {
+	treeID     int64
+	instanceID string
+	lockFile   string
+
+	client   *clientv3.Client
+	session  *concurrency.Session
+	election *concurrency.Election
+
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
+// Await blocks until the instance captures mastership. Returns the context
+// which remains active until the instance stops believing to be the master,
+// or the passed in context is canceled. Returns an error if capturing fails
+// or the passed in context is canceled before that.
+func (e *Election) Await(ctx context.Context) (context.Context, error) {
+	if e.ctx != nil && e.ctx.Err() == nil {
+		return e.ctx, nil
+	}
```

Comment:
I added this behavior to not spin up multiple `Observe` goroutines if `Await` is called multiple times intentionally or mistakenly called second time before any `Resign`. Also, we are storing the corresponding `cancel` anyway to make `Resign/Close` work. If we remove `ctx` but leave the `cancel` then misusing `Await/Resign` sequence will result in leakage (only the latest context/goroutine will be canceled on `Close`).

Reply:
Having `e.ctx!=nil` as a test for whether a goroutine is running seems obscure and racy -- wouldn't a `sync.WaitGroup` be a better tool for the job?

Reply:
Dunno. W.r.t. raciness: I did not intend it to be thread-safe anyway because that is a strong thing to require in an interface, plus etcd's `Election` does not seem to cope well with concurrent use.

I agree with the obscurity point, but maybe a comment would suffice?

How do you think `WaitGroup` should be used? Putting a `wg.Wait()` here also has a disadvantage: it can stuck forever if `Resign` was not called.

Reply:
An orthogonal question: Maybe a good idea is to just return an error here, so that the client would always have to `Resign`?

Reply:
Returning the current master context contradicts a little bit with how the passed in `ctx` should be treated.

Reply:
I had been thinking that you could add `wg sync.WaitGroup` to `Election` instead of `ctx` and have:
```go
	e.wg.Add(a)
	go func() {
		defer e.wg.Done()
		for rsp := range ch {
			...
```
then have `Resign` do `e.wg.Wait()` after invoking the cancel:
```go
func (e *Election) Resign(ctx context.Context) error {
	err := e.election.Resign(ctx)
	// Cancel after Resign to tolerate situations when ctx is the master context.
	if e.cancel != nil {
￼		e.cancel()
		e.cancel = nil
￼	}
	e.wg.Wait()
	return err
￼}
```

However, that wouldn't protect against your scenario of 2 x `Await` calls -- you'd still need some protection at the top of the `Await()` method like:
```go
func (e *Election) Await(ctx context.Context) (context.Context, error) {
	if e.cancel != nil {
		// Terminate existing monitoring goroutine
		e.cancel()
		e.cancel = nil
	}
	...
```
(which doesn't seem that much better/clearer than the starting point...)

Reply:
(A wait group might be overkill, but it does seem like you're storing a context just so you can treat `e.ctx != nil` as a boolean that indicates whether a monitoring goroutine is running...so why not just use have a `monitoringRunning bool` field instead?)

Reply:
I considered that as well. The first observation was that this `bool` was accessed from 2 goroutines: the tracking one, and the one that calls `Resign`. So it had to be an `int32` and read/written atomically. The second observation was that setting it to zero would happen exactly where `cancel()` is invoked, so basically it would mirror the effect of `cancel()`. I thought reading it directly from `ctx` might be simpler/shorter, as it conveniently covers both atomicity and the cancelation bit.

Reply:
Hm, maybe I could put zero assigning to end of the monitoring goroutine instead of next to each `cancel`. This is better because it would actually prevent 2 monitoring goroutines running simultaneously, whereas starting another goroutine when `ctx` is canceled does not yet guarantee the previous goroutine has stopped.

Reply:
Ah, but the latter idea is not great from the interface user's perspective. After they call `Resign`, they probably want to be able to `Await`, but since zero-setting the `bool` is deferred until the monitoring goroutine actually terminates, they will only be able to `Await` after indeterministic amount of time. So we get back to "zero-setting should happen next to `cancel()`".

Reply:
To have both a satisfactory interface behavior and not store `ctx` we could also use a channel, so that:
- `Await` checks if the channel is not nil and is open, and returns an error if so.
- Otherwise it stores a new `done` channel and runs the goroutine with `defer close(done)`.
- `Resign` waits for `<-done` after calling `cancel()`.

Reply:
But then again, there is a flaw. When the instance loses mastership and closes mastership context, `Await` will start working after an indeterminate amount of time (might take a while to close `done`). So either we should also export a `func Done() chan struct{}` in the interface to give the client the ability to wait for it, or get back to `done==canceled` behavior.

Reply:
Sorry for a lot of comments...

The latter flaw can be avoided if we re-order defers a bit:
```

go func() {
  defer cancel()
  defer close(done)
  ...
}()
```
This way, when the mastership context is canceled forcefully, it will automatically imply `done`, and thus `Await` can be called immediately.

Reply:
Please have a look at the new code.